### ISSUE TITLE
State ceph.rgw restarts radosgw without config change

### ISSUE THREAD

**Issue by mkoderer** [SUSE/DeepSea#584]
The state ceph.rgw restarts the service although no config changes were done.
This can be very dangerous in productions since it could lead to a short outage even though rgw is deployed multiple times.

By using:
 salt -C 'I@roles:rgw' state.apply ceph.rgw
It can happen that all rgw instances are restarted.

Safe workaround:
 salt -C 'I@roles:rgw' -b 1 --failhard state.apply ceph.rgw

It would be much better to use the `watch` function in salt to monitor changes in config variables.

**Comment by jan--f** [comment]
@mkoderer good point yes. Adding a ```watch``` statement would surely fix the issue for no config changes. On the other hand if the config on all rgw nodes changed the restart of all nodes might still overlap.
Would you see this as an expected downtime? Otherwise we could add some code to serialise thses daemon restarts, making sure that not all get restarted in parallel.

**Comment by mkoderer** [comment]
As long a test=True indicates that there is a config change + service restart it's fine.
I am not aware of any function that could serialise a service restart..

**Comment by jan--f** [comment]
That would not be a state based solution. We have done this before for stage 0 (see https://github.com/SUSE/DeepSea/blob/master/srv/salt/ceph/stage/prep/minion/default.sls#L33). It involved iterating over the results of a runner...not terribly complicated. The question is how big (if any) the benefit of this is.

**Comment by swiftgist** [comment]
We did serialize steps in Stage 0 when a Ceph cluster is already running.  I expect we need to do this for each of the services as well.  (That is, fresh install should do parallel and existing install should do things serially.)

**Comment by mkoderer** [comment]
Ah that's nice - I wasn't aware of this feature.
What would be also very useful is the "--failhard" behavior. Means if first restart didn't work stop restarting. Otherwise an incorrect setting can lead to a complete outage.

But anyway - if we can start with "watch" and restarting only when needed it would be sufficient as first step IMO.

**Comment by swiftgist** [comment]
Just for reference, issue #221 has been on our todo list.

**Comment by jschmid1** [comment]
A quick update on our findings:

### Preface:

The way we do config changes is (for technical reasons) staged. There is:

`salt -C 'roles:master' state.apply ceph.configuration.create` 

which collects any files under `/srv/salt/ceph/configuration/files/ceph.conf.d`

(and `/srv/salt/ceph/configuration/files/ceph.conf.*` -- that be adapted to follow the design above)

and creates a ceph.conf that is temporarily saved under `/srv/salt/ceph/configuration/cache/ceph.conf:`

`salt '*' state.apply ceph.configuration` then distributes the written config to the minions.

These two states are currently being executed within stage.3 
In order to limit a rgw restart to the event of a changed config file we evaluated several options:

- *1*: Use salt's [watch](https://docs.saltstack.com/en/2015.8/ref/states/requisites.html)

Salt's watch requisite requires a state or a module that can be checked for it's return value.
I.e. in our case we need to pull in the ceph.conf creation to the place where we invoke the `service.running` module

```
include:
  - ..configuration

service.running:
  name: rgw
    - watch:
      - file: /etc/ceph/conf
```

As `ceph.configuration` is only _distributing_ the config, it needs to be _created_ before, requiring to also include ceph.configuration.create, which does not work inside this state file.

Which basically leaves us with the twisted workflow of.
1) Adapt the ceph.conf.d/X.conf
2) Run ceph.configuration.create
3) Trigger Stage.4

That leads to an additional -- the presence of `configuration.create` and `configuration` in stage.3.
If you happen to run stage.3 before stage.4 you loose your 'success_event' that `watch` is getting from the `configuration` event, hence you will not trigger a restart although there was a config change.

- *2*: Use a beacon's [inotify](https://docs.saltstack.com/en/latest/ref/beacons/all/salt.beacons.inotify.html#module-salt.beacons.inotify)

Salt's beacons module for inotify can monitor specified files. This would enable us to globally track changes in files. Unfortunately, a beacon can only send event's to the bus. To 'react' to those events you need to listen with a [reactor](https://docs.saltstack.com/en/latest/topics/reactor/).
This would mean that any change to a ceph.conf file will trigger a restart of a given service. No simple solutions of adding a delay or even a 'planned maintenance window'. You can get notified of course. But this do not solve the initial problem.


- *3*: Use a custom 'config_has_changed' mechanism

 This will be yet another runner that detects changes in the `ceph.conf.d/sub_files.conf` and compares those to temp files based on their timestamps. A problem which I didn't mention earlier is the nature of the ceph.conf itself. It combines all config settings for any service which makes it very hard to determine which service received the change in settings.
Using this approach we can listen on the files in `ceph.conf.d` and react accordingly in the 'service.restart' states.

Starting with he implementation of **3**.

HTH

**Comment by swiftgist** [comment]
The fix is in #632
